### ISSUE TITLE
Cannot restore non-zipped backup on Windows (5.0)

### ISSUE THREAD

**Issue by merenyics** [eXist-db/exist#3026]
### What is the problem
**Update: Zip archives can be restored, but if you need to create your own zip file, be sure to use `jar.exe` to pack the backup, otherwise the zip file may be unreadable by exist. This may be worth documenting. Non-zipped backups continue to fail on Windows, this needs to be fixed.**

Trying to restore a non-zipped (full) backup fails with an error message saying that some parts of the temporary file created in windows/temp do not exist or cannot be read. The parts listed in the error message (windows/temp/....tmp#...) are basically all the collections directly below /db.

The temp file mentioned in the error message seems to be a zip file at first glance, but after renaming it to *.zip, I cannot open it as a zip archive. Is this normal?

I also had other problems with restoring a zip archive:
 - I created a zip backup (this one could be restored, except some malformed files). 
 - I unzipped the contents and changed some files to fix some well-formedness errors inside xml comments.
- I zipped the entire collection again, and tried to restore the zipped archive. 
- Restore failed with exception, saying something (not sure what) is MALFORMED.

### What did you expect
A restored backup.

### Describe how to reproduce or add a test
I'm afraid the original backup I'm having a problem with is confidential, but let's hope that simply trying to restore any non-zipped backup on a windows server will reproduce the issue.

### Context information
Please always add the following information
- eXist-db version 5.0 release
-  java version "1.8.0_161" Java(TM) SE Runtime Environment (build 1.8.0_161-b12)
- Operating system Windows 10
- 64 bit
- How is eXist-db installed? jar
- no custom changes

**Comment by craigberry** [comment]
Check your [storage sense](https://techcommunity.microsoft.com/t5/Storage-at-Microsoft/Windows-10-and-Storage-Sense/ba-p/428270) settings.  Try doing a restore with storage sense turned off and see if that makes a difference.  It might be hard to reproduce because it might depend on how close you are to filling up the disk.

I can't prove it but I have a strong suspicion that the most robust path forward for eXist will be, when it creates a temp copy of the backup on the server, to create a zip backup regardless of whether the original backup was a zip or a directory tree.  That way the zip file will remain open as long as the restore is in progress and will presumably be left alone by a temp file cleaner.  As a side effect, it will put less pressure on the server disk space, which is also a benefit.

N.B.  This does not seem to be a Windows-only problem as there was a report on the [mailing list](https://sourceforge.net/p/exist/mailman/message/36756468/) by someone using Linux.

**Comment by merenyics** [comment]
@craigberry  Thanks for the reply. Unfortunately, my problem looks different from the one described in the mailing list. The temp files are still present in the temp folder (and the disk is almost empty).

Can anyone confirm that restoring non-zipped backups works in 5.0? Also, can anyone shed some light on why zipping a backup manually (using the internal zip packer in total commander) and trying to restore that zipped backup might fail?

**Comment by merenyics** [comment]
this time I managed to copy the error message from the java admin client, I hope this helps:
```
org.xmldb.api.base.XMLDBException: 
	at org.exist.xmldb.RemoteRestoreService.restore(RemoteRestoreService.java:135)
	at org.exist.client.ClientFrame.lambda$39(ClientFrame.java:1122)
	at java.lang.Thread.run(Unknown Source)
Caused by: org.apache.xmlrpc.XmlRpcException: Failed to invoke method getRestoreTaskEvents in class org.exist.xmlrpc.RpcConnection: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: MALFORMED
	at org.apache.xmlrpc.client.XmlRpcStreamTransport.readResponse(XmlRpcStreamTransport.java:197)
	at org.apache.xmlrpc.client.XmlRpcStreamTransport.sendRequest(XmlRpcStreamTransport.java:156)
	at org.apache.xmlrpc.client.XmlRpcHttpTransport.sendRequest(XmlRpcHttpTransport.java:143)
	at org.apache.xmlrpc.client.XmlRpcSunHttpTransport.sendRequest(XmlRpcSunHttpTransport.java:69)
	at org.apache.xmlrpc.client.XmlRpcClientWorker.execute(XmlRpcClientWorker.java:56)
	at org.apache.xmlrpc.client.XmlRpcClient.execute(XmlRpcClient.java:167)
	at org.apache.xmlrpc.client.XmlRpcClient.execute(XmlRpcClient.java:158)
	at org.apache.xmlrpc.client.XmlRpcClient.execute(XmlRpcClient.java:147)
	at org.exist.xmldb.RemoteRestoreService.restore(RemoteRestoreService.java:128)
	... 2 more

```

**Comment by dizzzz** [comment]
DO you use client-side and server side the same version of the database?

**Comment by merenyics** [comment]
i'm using the java client on the server, which is v5.0, so there is no mismatch between the java admin client and the server, to the best of my knowledge.
The backup itself comes from 4.x, I'm trying to migrate my data to the new version.

**Comment by merenyics** [comment]
there is very little evidence in the log file of what went wrong, but this seems important:
`2019-09-25 19:47:44,707 [db.exist.rpc-db-restore-2] WARN  (TemporaryFileManager.java [returnTemporaryFile]:133) - Unable to delete temporary file: C:\WINDOWS\TEMP\exist-db-temp-file-manager-5601779988568748619\exist-db-temp-810877875223003118.tmp due to: C:\WINDOWS\TEMP\exist-db-temp-file-manager-5601779988568748619\exist-db-temp-810877875223003118.tmp: A folyamat nem fér hozzá a fájlhoz, mert azt egy másik folyamat használja.`
The error message at the end is in Hungarian, it means: "the process cannot access the file, because it is being used by another process."

This time the .tmp file was a perfectly valid .zip file, which I was able to unpack using any tool, and it contained the entire backup.

**Comment by merenyics** [comment]
...still unable to restore a backup unless it is an intact zip backup.

Can anyone familiar with the internals of the restore process shed some light on what could go wrong if the backup contents are changed in any way?  I'd especially like to know:

- Should timestamps in the `__contents__` files be updated (to match the actual timestamp on the files themselves) after some xml documents are changed?
-  Is there any restriction on the (type/version of) zip archiver that can be used?

I can think of no other reason why my backups should be any different from the ones created directly by exist. Any suggestion would be much appreciated.

**Comment by merenyics** [comment]
I tried to test all possible scenarios, and found the following:


- restoring an intact zip archive works OK
- on my windows PC, restoring a non-zipped backup (whether it's been tampered with or not) simply doesn't work.
- re-zipping a **perfectly intact** backup (one that is tested to work), then trying to restore that results in the following exception being thrown

```
Caused by: java.lang.IllegalArgumentException: MALFORMED
	at java.util.zip.ZipCoder.toString(Unknown Source) ~[?:1.8.0_221]
	at java.util.zip.ZipFile.getZipEntry(Unknown Source) ~[?:1.8.0_221]
	at java.util.zip.ZipFile.access$900(Unknown Source) ~[?:1.8.0_221]
	at java.util.zip.ZipFile$ZipEntryIterator.next(Unknown Source) ~[?:1.8.0_221]
	at java.util.zip.ZipFile$ZipEntryIterator.nextElement(Unknown Source) ~[?:1.8.0_221]
	at java.util.zip.ZipFile$ZipEntryIterator.nextElement(Unknown Source) ~[?:1.8.0_221]
	at org.exist.backup.ZipArchiveBackupDescriptor.countFileEntries(ZipArchiveBackupDescriptor.java:239) ~[exist-core-5.0.0.jar:5.0.0]
	at org.exist.backup.ZipArchiveBackupDescriptor.<init>(ZipArchiveBackupDescriptor.java:57) ~[exist-core-5.0.0.jar:5.0.0]
	at org.exist.backup.Restore.getBackupDescriptor(Restore.java:157) ~[exist-core-5.0.0.jar:5.0.0]
	at org.exist.backup.Restore.getBackupDescriptors(Restore.java:116) ~[exist-core-5.0.0.jar:5.0.0]
	at org.exist.backup.Restore.restore(Restore.java:68) ~[exist-core-5.0.0.jar:5.0.0]
	at org.exist.xmlrpc.RpcConnection.lambda$124(RpcConnection.java:3671) ~[exist-core-5.0.0.jar:5.0.0]
	at org.exist.xmlrpc.function.XmlRpcFunction.apply(XmlRpcFunction.java:45) ~[exist-core-5.0.0.jar:5.0.0]
	at org.exist.xmlrpc.RpcConnection.withDb(RpcConnection.java:4061) ~[exist-core-5.0.0.jar:5.0.0]
	at org.exist.xmlrpc.RpcConnection.withDb(RpcConnection.java:4045) ~[exist-core-5.0.0.jar:5.0.0]
	at org.exist.xmlrpc.RpcConnection.lambda$123(RpcConnection.java:3669) ~[exist-core-5.0.0.jar:5.0.0]
	at java.util.concurrent.FutureTask.run(Unknown Source) ~[?:1.8.0_221]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:1.8.0_221]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:1.8.0_221]
	... 1 more
```
my conclusion: 

1. the actual zip algorithm used to create the backup file does matter, so what should I use on windows?
2. there is a problem with the temporary zip created while restoring a non-zipped backup

This is all on a windows server, other OSs might work.

This can be a bit of a headache, as the backups can end up being invalid because of serialization bugs present in 4.x. I can fix the broken files, but then I have no way of restoring the db. My last resort is rebuilding the db from the original files piece by piece - not ideal.

**Comment by dizzzz** [comment]
> the actual zip algorithm used to create the backup file does matter, so what should I use on windows?

Yes, the ZIP specification is a container specification, I remember from a long time that java does not support all options here. If you use the java JAR.exe, things should work OK.

**Comment by merenyics** [comment]
Seems to have been solved, see #3196
